File: Ruby/napakalaki/lib/numeric_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'bad_consequence.rb'

module NapakalakiGame

class NumericBadConsequence < BadConsequence
  
  def initialize(aText, nLevels, nVisibleTreasures, nHiddenTreasures)
    
    super(aText, nLevels)
    @nVisibleTreasures = nVisibleTreasures
    @nHiddenTreasures = nHiddenTreasures
    
  end
  
  public_class_method :new
    
  
  def empty?
    @nVisibleTreasures == 0 && @nHiddenTreasures == 0
  end 
  
  def to_s
    "\n\tText: #{@aText} \n\tLevels lost: #{@nLevels} \n\tVisible treasures lost: #{@nVisibleTreasures} \n\tHidden treasures lost: #{@nHiddenTreasures} \n\tDeath: #{@death}"
  end
  
  def substractVisibleTreasure(t)
     @nVisibleTreasures -= 1 if @nVisibleTreasures > 0
        
  end
  
  def substractHiddenTreasure(t)
    @nHiddenTreasures -= 1 if @nHiddenTreasures > 0
        
  end
  
  
  def adjustToFitTreasureList(vTreasures, hTreasures)
            
    nvisible = [vTreasures.size, @nVisibleTreasures].min
    nhidden = [hTreasures.size, @nHiddenTreasures].min
                    
    return NumericBadConsequence.new(@aText, 0, nvisible, nhidden)

  end
  
end
end

File: Java/src/napakalaki/NumericBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;

/**
 *
 * @author Ludvins
 */
public class NumericBadConsequence extends BadConsequence{
    
    protected int nVisibleTreasures;
    protected int nHiddenTreasures;
    
    public NumericBadConsequence(String text, int levels, int nVisible, int nHidden){
        this.text = text;
        this.levels = levels;
        this.nVisibleTreasures = nVisible;
        this.nHiddenTreasures = nHidden;
        this.death = false;
    }
    
    @Override
    public boolean isEmpty(){
       return this.nHiddenTreasures == 0 && this.nVisibleTreasures == 0;
    }
    
    @Override
    public NumericBadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h){
        
        return new NumericBadConsequence(text, 0, Math.min(nVisibleTreasures, v.size()), Math.min(nHiddenTreasures, h.size()));
         
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){
        if (this.nVisibleTreasures > 0) this.nVisibleTreasures--;
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t){
        if (this.nHiddenTreasures > 0 ) this.nHiddenTreasures--;
    }
    
    public int getnVisibleTreasures() {
        return nVisibleTreasures;
    }

    public int getnHiddenTreasures() {
        return nHiddenTreasures;
    }

    @Override
    public String toString() {
        return "\n\tText: " + text + "\n\tLevels: " + levels + "\n\tnVisibleTreasures: " + nVisibleTreasures + "\n\tnHiddenTreasures: " + nHiddenTreasures + "\n\tDeath: " + death;
    }
}
